index/scorch/snapshot_index.go

```diff
@@ -43,6 +44,29 @@ type IndexSnapshot struct {
 	offsets  []uint64
 	internal map[string][]byte
 	epoch    uint64
+
+	refs int64 // Access only through sync/atomic.
+}
+
+func (i *IndexSnapshot) AddRef() {
+	atomic.AddInt64(&i.refs, 1)
+}
+
+func (i *IndexSnapshot) DecRef() (err error) {
+	if i == nil || atomic.AddInt64(&i.refs, -1) != 0 {
```

Comment:
generally we never check if the receiver is nil, we assume is programmer error to have not checked the return value when creating the index.